register-machines/src/main.rs

```diff
@@ -13,66 +15,165 @@ enum Instruction {
 
 use Instruction::*;
 fn eval_program(program: &[Instruction], state: &State) -> State {
-    unimplemented!();
+    let (mut l, mut rs) = state.clone();
+    while l < program.len() {
+        match program[l] {
+            Add(i, j) => {
+                *rs.entry(i).or_insert(Godel::new()) += 1;
+                l = j; 
+            },
+            Sub(i, j, k) => {
+                let v = rs.entry(i).or_insert(Godel::new()); 
+                if *v != 0 {
+                    *v -= 1;
+                    l = j;
+                } else {
+                    l = k; 
+                }
+            },
+            Halt => { break; }
+        }
+    }
+    (l, rs)
+}
```

Comment:
Great stuff. Looks like the reference solution and compiles into some very nice and efficient code :)